frontend/src/metabase/visualizations/components/Visualization/Visualization.jsx

```diff
@@ -72,39 +72,87 @@ const mapStateToProps = state => ({
 
 const SMALL_CARD_WIDTH_THRESHOLD = 150;
 
-class Visualization extends PureComponent {
-  state = {
-    getHref: undefined,
-    hovered: null,
-    clicked: null,
-    error: null,
-    genericError: null,
-    warnings: [],
-    series: null,
-    visualization: null,
-    computedSettings: {},
+const isLoading = series => {
+  return !(
+    series &&
+    series.length > 0 &&
+    _.every(
+      series,
+      s => s.data || _.isObject(s.card.visualization_settings.virtual_card),
+    )
+  );
+};
+
+const deriveStateFromProps = props => {
+  const transformed = props.rawSeries
+    ? getVisualizationTransformed(extractRemappings(props.rawSeries))
+    : null;
+
+  const series = transformed?.series;
+
+  const computedSettings = !isLoading(series)
+    ? getComputedSettingsForSeries(series)
+    : {};
+
+  return {
+    series,
+    computedSettings,
+    visualization: transformed?.visualization,
   };
+};
+
+class Visualization extends PureComponent {
+  constructor(props) {
+    super(props);
 
-  UNSAFE_componentWillMount() {
-    this.transform(this.props);
+    this.state = {
+      getHref: undefined,
+      hovered: null,
+      clicked: null,
+      error: null,
+      genericError: null,
+      warnings: [],
+      series: null,
+      visualization: null,
+      computedSettings: {},
+      ...deriveStateFromProps(props),
+    };
   }
 
-  UNSAFE_componentWillReceiveProps(newProps) {
+  static getDerivedStateFromProps(props, state) {
```

Comment:
Using `componentDidUpdate` caused 2 tests to fail, one of which is `should show user-friendly error when combining series that cannot be visualized together (metabase#32231)` and it is a genuine failure. It seems to be caused by the slight behavior difference (before vs after update) in React. I found that the only possible solution is to use `getDerivedStateFromProps` for state updates.